src/rust/fgbutil/src/main.rs

```diff
@@ -34,13 +38,54 @@ enum Format {
     Geojson,
 }
 
+// TODO: Cannot pass around GeozeroDatasource because it is a non object safe trait?
+// TODO: Regardless might need dyn / Box?
+
+fn write(format: Format, reader: GeozeroDatasource, output: BufWriter<File>) -> Result<()> {
```

Comment:
> Now I have another question.. I'd like to understand how to break about the main into methods but the outcommented atempt doesn not work because "GeozeroDatasource" is not object safe. I would like to know more about this and how I can break out the code...

tldr;
```
fn write(format: Format, reader: impl GeozeroDatasource, output: BufWriter<File>) -> Result<()> {
```

Longer answer (someone please correct or add to this if I'm wrong):

As you probably know, GeozeroDatasource is a trait. It can't be instantiated on its own. Instead, it's behavior that can be ascribed to potentially many different instantiable things. Those things are said to "implement GeozeroDatasource".

A "trait objects" is a pointer that could point to any thing that implements the trait. It's spelled like `Box<dyn GeozeroDatasource>`. 

As I understand it, this format is similar to how most other languages pass around trait-like things (e.g. arguments as protocols in objective-c/swift, or interfaces in java).

There are implications to this, rust errs on the side of being explicit, which is why it's such a mouthful to spell in rust. 

So in situations like this you *could* use a trait object, but only "object safe" traits can be made into a trait object. 

But what does "object safe" mean?
See https://doc.rust-lang.org/book/ch17-02-trait-objects.html#object-safety-is-required-for-trait-objects for the gospel.

But, since I know you write some java, a more intuitive example might be to read this persons question: https://stackoverflow.com/questions/8161712/how-can-i-refer-to-the-class-type-a-interface-is-implementing-in-java

I think that's philosophically similar issue - there are some methods defined in a trait that would need to *know* what the underlying implementing type is.

Since "trait objects" aren't usable in this case, another option we have to say "this argument could be one of  potentially different types that implement `GeoZeroDatasource`", is to use generics to compile a version of `write` for every underlying type that ultimately uses it.

```
fn write<G: GeozeroDatasource >(format: Format, reader: G, output: BufWriter<File>) -> Result<()> {
```

There is the "impl trait" sugar to equivalently spell this:

```
fn write(format: Format, reader: impl GeozeroDatasource, output: BufWriter<File>) -> Result<()> {
```

When both trait objects or generics are an option, there are pros and cons to either. (primarily: generics can perform better, but because generics produce multiple versions of the function,  it's slower to compile and bloats code size).

Reply:
Thanks @michaelkirk, it makes sense and explains why the changes suggested by @pka works. I still have issues internalizing the trait/impl part of the Rust type system but at least this is a step in the right direction. :)